vscode/src/language-service/testExplorer.ts

```diff
@@ -0,0 +1,219 @@
+    testCase: vscode.TestItem,
+    request: vscode.TestRunRequest,
+    worker: ICompilerWorker,
+  ): Promise<void> {
+    log.trace("Running Q# test: ", testCase.id);
+    if (testCase.children.size > 0) {
+      for (const childTestCase of testCase.children) {
+        await runTestCase(ctrl, childTestCase[1], request, worker);
+      }
+      return;
+    }
+    const run = ctrl.createTestRun(request);
+    const evtTarget = createDebugConsoleEventTarget((msg) => {
+      run.appendOutput(`${msg}\n`);
+    });
+    evtTarget.addEventListener("Result", (msg) => {
+      if (msg.detail.success) {
+        run.passed(testCase);
+      } else {
+        const failureLocation =
+          msg.detail?.value?.uri ||
+          (msg.detail?.value?.related &&
+            msg.detail.value.related[0].location?.source) ||
+          null;
+
+        const message: vscode.TestMessage = {
+          message: msg.detail.value.message,
+          location:
+            failureLocation === null
+              ? undefined
+              : {
+                  range: toVsCodeRange(msg.detail.value.range),
+                  uri: vscode.Uri.parse(failureLocation),
+                },
+        };
+        run.failed(testCase, message);
+      }
+      run.end();
+    });
+
+    const callableExpr = `${testCase.id}()`;
+    const uri = testCase.uri;
+    if (!uri) {
+      log.error(`No compilation URI for test ${testCase.id}`);
+      run.appendOutput(`No compilation URI for test ${testCase.id}\r\n`);
+      return;
+    }
+    const programResult = await getProgramForDocument(uri);
+
+    if (!programResult.success) {
+      throw new Error(programResult.errorMsg);
+    }
+
+    const program = programResult.programConfig;
+
+    try {
+      await worker.run(program, callableExpr, 1, evtTarget);
+    } catch (error) {
+      log.error(`Error running test ${testCase.id}:`, error);
+      run.appendOutput(`Error running test ${testCase.id}: ${error}\r\n`);
+    }
+    log.trace("ran test:", testCase.id);
+  }
+
+  testController.createRunProfile(
+    "Interpreter",
+    vscode.TestRunProfileKind.Run,
+    runHandler,
+    true,
+    undefined,
+    false,
+  );
+
+  const testVersions = new WeakMap<vscode.TestItem, number>();
+  async function onTestCallables(evt: {
+    detail: {
+      callables: ITestDescriptor[];
+    };
+  }) {
+    let currentVersion = 0;
+    for (const [, testItem] of testController.items) {
+      currentVersion = (testVersions.get(testItem) || 0) + 1;
+      break;
+    }
+
+    for (const { callableName, location, friendlyName } of evt.detail
+      .callables) {
+      const vscLocation = toVsCodeLocation(location);
+      const parts = [friendlyName, ...callableName.split(".")];
+
+      let rover = testController.items;
+      for (let i = 0; i < parts.length; i++) {
+        const part = parts[i];
+        const id =
+          i === parts.length - 1
+            ? callableName
+            : parts.slice(0, i + 1).join(".");
```

Comment:
I'm struggling to follow the logic here a bit. Can you add some explanatory comments? For example, parts is made up of the callable name split by ".", and here you are iterative over those parts, yet if on the last part the 'id' is the whole callable name, not including the friendly name and not split by ".", but if on an earlier part it is including the friendlyName and upto and including that part?

I'm not sure I follow. Some comments would be a lot easier than having to reverse engineer to try and understand.

Reply:
I've documented this in 92fbb1a -- let me know if it still isn't clear